game-core/src/main/java/games/strategy/triplea/delegate/RocketsFireHelper.java

```diff
@@ -68,64 +122,96 @@ private static boolean isLimitRocketDamageToProduction(final GameData data) {
     return Properties.getLimitRocketAndSbrDamageToProduction(data);
   }
 
-  static void fireRockets(final IDelegateBridge bridge, final PlayerId player) {
-    final GameData data = bridge.getData();
-    final Set<Territory> rocketTerritories = getTerritoriesWithRockets(data, player);
-    if (rocketTerritories.isEmpty()) {
-      bridge.getHistoryWriter().startEvent(player.getName() + " has no rockets to fire");
-      return;
-    }
-    if (isWW2V2(data) || isAllRocketsAttack(data)) {
-      fireWW2V2(bridge, player, rocketTerritories);
-    } else {
-      fireWW2V1(bridge, player, rocketTerritories);
-    }
-  }
-
-  private static void fireWW2V2(
-      final IDelegateBridge bridge,
-      final PlayerId player,
-      final Set<Territory> rocketTerritories) {
+  /**
+   *  Find Rocket Targets and load up fire rockets for later execution if necessary.
+   *  Directly fired with Sequentially Targeted rockets.
+   */
+  private void findRocketTargetsAndFireIfNeeded(final IDelegateBridge bridge, final boolean fireRocketsImmediately) {
     final GameData data = bridge.getData();
-    final Set<Territory> attackedTerritories = new HashSet<>();
-    final boolean oneAttackPerTerritory = !isRocketAttacksPerFactoryInfinite(data);
-    for (final Territory territory : rocketTerritories) {
-      final Set<Territory> targets = getTargetsWithinRange(territory, data, player);
-      if (oneAttackPerTerritory) {
-        targets.removeAll(attackedTerritories);
+    final PlayerId player = bridge.getPlayerId();
+    final Map<Territory, Integer> previouslyAttackedTerritories = new LinkedHashMap<>();
+    for (final Territory attackFrom : getTerritoriesWithRockets(data, player)) {
+      final Set<Territory> targets = getTargetsWithinRange(attackFrom, data, player);
+      final int maxAttacks = TechAbilityAttachment.getRocketNumberPerTerritory(player, data);
+      for (final Territory t : previouslyAttackedTerritories.keySet()) {
+        // negative Rocket Number per Territory == unlimited
+        if (maxAttacks < 0 || maxAttacks <= previouslyAttackedTerritories.get(t).intValue()) {
```

Comment:
If maxAttacks == -1 then shouldn't we not remove this from targets since its unlimited?

Reply:
You're quite correct! I had maxAttacks set to 200 in Global so I didn't spot this bug.